037-prototypes/proposal.md

```diff
@@ -0,0 +1,467 @@
+  to care where configuration is being set in pipelines:
+  [concourse/git-resource#172](https://github.com/concourse/git-resource/pull/172),
+  [concourse/bosh-deployment-resource#13](https://github.com/concourse/bosh-deployment-resource/issues/13),
+  [concourse/bosh-deployment-resource#6](https://github.com/concourse/bosh-deployment-resource/issues/6),
+  [concourse/cf-resource#20](https://github.com/concourse/cf-resource/pull/20),
+  [concourse/cf-resource#25](https://github.com/concourse/cf-resource/pull/25),
+  [concourse/git-resource#210](https://github.com/concourse/git-resource/pull/210)
+
+* Make resource actions reentrant so that we no longer receive `unexpected EOF`
+  errors when reattaching to an in-flight build whose resource action completed
+  while we weren't attached:
+  [concourse/concourse#1580](https://github.com/concourse/concourse/issues/1580)
+
+* Support for showing icons for resources in the web UI:
+  [concourse/concourse#788](https://github.com/concourse/concourse/issues/788),
+  [concourse/concourse#3220](https://github.com/concourse/concourse/pull/3220),
+  [concourse/concourse#3581](https://github.com/concourse/concourse/pull/3581)
+
+* Standardize TLS configuration so every resource doesn't implement their own
+  way: [concourse/rfcs#9](https://github.com/concourse/rfcs/issues/9)
+
+
+## Glossary
+
+* **Prototype**: an implementation of the interface defined by this proposal,
+  typically packaged as an OCI container image.
+
+  Examples:
+
+  * a `git` prototype for interacting with commits and branches in a repo.
+  * a `time` prototype for performing interval triggers.
+
+* **Message**: an operation to perform against the object. Corresponds to a
+  command to run in the container image.
+
+  Examples:
+
+  * `check`
+  * `get`
+
+* **Object**: a logical entity, encoded as a JSON object, acted on and emitted
+  by a prototype in response to messages.
+
+  Examples:
+
+  * `{"uri":"https://github.com/concourse/rfcs"}`
+  * `{"ref":"e4be0b367d7bd34580f4842dd09e7b59b6097b25"}`
+
+* **Metadata**: additional information about an object to surface to users.
+
+  Examples:
+
+  * `[{"name":"committer","value":"Alex Suraci"}]`
+
+* **Bits**: a directory containing arbitrary data.
+
+  Examples:
+  
+  * source code checked out from a repo
+  * compiled artifacts created by a task or another prototype
+
+* **Resource**: a **prototype** which implements the following messages:
+
+  * `check`: discover objects representing versions, in order
+  * `get`: fetch an object version
+
+
+## Interface Types
+
+```go
+// Object is an object receiving messages and being emitted in response to
+// messages.
+type Object map[string]interface{}
+
+// InfoRequest is the payload written to stdin for the `./info` script.
+type InfoRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
```

Comment:
I'm on the fence about including this in `Info`. There may be merit in forbidding network activity in the `Info` call outright. I kind of just threw it in for symmetry.

Reply:
I suppose this could be useful for verification and for inspecting a remote entity to discover which messages it supports via the prototype. :thinking: 

...but that could also be problematic, as there would be confusion over whether the prototype even implements a message vs. whether the user has permission for the message, and I wouldn't expect all prototype authors to implement 'dynamic message support detection'.

So I'm still leaning towards just removing this and enforcing that `Info` is for local static processing and does not reach out to the network.